File: scripts/populate_posters.py
#!/usr/bin/env python3
# Generate media entries with poster URLs from public APIs.
# - Anime -> AniList GraphQL
# - Movies/Series/Music -> iTunes Search API
import json, os, re, sys
from urllib.request import Request, urlopen
from urllib.parse import quote_plus

TMDB_BASE = "https://image.tmdb.org/t/p/w780"
ANILIST_URL = "https://graphql.anilist.co"
ITUNES_URL = "https://itunes.apple.com/search"

HEADERS = {
    "User-Agent": "Mozilla/5.0 (compatible; vault-populate/1.0)",
    "Accept": "application/json",
}

def fetch_json(url, data=None, headers=None):
    req = Request(url, data=data, headers=headers or HEADERS, method="POST" if data else "GET")
    with urlopen(req, timeout=25) as r:
        return json.loads(r.read().decode("utf-8"))

def anilist_poster(title, year=None):
    query = """
    query ($search: String) {
      Page(page: 1, perPage: 5) {
        media(search: $search, type: ANIME, sort: POPULARITY_DESC) {
          title { romaji english native }
          coverImage { extraLarge large }
          startDate { year }
          siteUrl
        }
      }
    }
    """
    try:
        payload = json.dumps({"query": query, "variables": {"search": title}}).encode()
        data = fetch_json(ANILIST_URL, data=payload, headers={**HEADERS, "Content-Type": "application/json"})
        media = (((data.get("data") or {}).get("Page") or {}).get("media") or [None])[0]
        if not media:
            return None
        yr = ((media.get("startDate") or {}).get("year")) or year
        cover = (((media.get("coverImage") or {}).get("extraLarge")) or ((media.get("coverImage") or {}).get("large")))
        t = ((media.get("title") or {}).get("english") or (media.get("title") or {}).get("romaji") or title)
        if not cover:
            return None
        return {"title": t, "year": yr, "poster": cover, "source": "anilist"}
    except Exception as e:
        print(f"[anilist] {title}: {e}", file=sys.stderr)
        return None

def itunes_artwork(title, media="all", year=None):
    try:
        params = f"?term={quote_plus(title)}&media={media}&limit=5&country=US"
        data = fetch_json(ITUNES_URL + params)
        results = data.get("results") or []
        if not results:
            return None
        best = results[0]
        best_art = (best.get("artworkUrl100") or "").replace("100x100", "600x600")
        for r in results:
            art = (r.get("artworkUrl100") or "").replace("100x100", "600x600")
            if art and len(art) > len(best_art):
                best = r
                best_art = art
        if not best_art:
            return None
        t = best.get("collectionName") or best.get("trackName") or title
        yr = (best.get("releaseDate") or "").split("-")[0] or year
        return {"title": t, "year": int(yr) if yr and yr.isdigit() else year, "poster": best_art, "source": "itunes"}
    except Exception as e:
        print(f"[itunes] {title}: {e}", file=sys.stderr)
        return None

def slugify(text):
    text = text.lower()
    text = re.sub(r"[^a-z0-9]+", "-", text)
    return text.strip("-")

series = [
    ("Spartacus", "series", 2010, ["Drama","Acción"], "Serie de gladiadores y rebelión contra Roma; blood, sand and steel."),
    ("The Last of Us", "series", 2023, ["Drama","Post-apocalíptico"], "Joel y Ellie en un mundo devastado por hongos."),
    ("Peaky Blinders", "series", 2013, ["Drama","Crimen"], "Familia Shelby domina Birmingham entre guerras y negocios oscuros."),
    ("Euphoria", "series", 2019, ["Drama"], "Adolescencia sin filtro en torno a Zendaya."),
    ("True Detective", "series", 2014, ["Drama","Crimen"], "Temporadas tipo antología con casos oscuros y personajes atormentados."),
    ("The Queen's Gambit", "series", 2020, ["Drama"], "Ajedrez y obsesión en los años 60."),
    ("Sopranos", "series", 1999, ["Drama","Crimen"], "Terapia, familia y mafia en Nueva Jersey."),
    ("The Walking Dead", "series", 2010, ["Drama","Terror"], "Supervivencia extrema tras el apocalipsis zombi."),
    ("Dexter", "series", 2006, ["Drama","Crimen"], "Forense de día, justiciero nocturno de asesinos."),
    ("Mr. Robot", "series", 2015, ["Tecnología","Thriller"], "Hacker, ansiedad y hackeo global."),
    ("The Boys", "series", 2019, ["Acción","Comedia"], "Superhéroes corruptos versus humanos valientes."),
]

anime = [
    ("Fullmetal Alchemist: Brotherhood", "anime", 2009, ["Aventura","Fantasía"], "Alquimia, hermanos Elric y cero relleno."),
    ("Tokyo Ghoul", "anime", 2014, ["Terror","Drama"], "Entre humanos y ghouls con dilema de identidad."),
    ("Vinland Saga", "anime", 2019, ["Acción","Histórico"], "Vikingos, venganza y redención."),
    ("Sword Art Online", "anime", 2012, ["Ciencia ficción"], "MMORPG mortal entre la línea real y virtual."),
    ("Rick and Morty", "anime", 2013, ["Comedia","Ciencia ficción"], "Viajes dimensionales con humor negro."),
    ("Hunter x Hunter", "anime", 2011, ["Aventura","Fantasía"], "Exámenes Hunter, bestias cardenales y amistad."),
    ("Cowboy Bebop", "anime", 1998, ["Acción","Ciencia ficción"], "Cazarrecompensas espaciales con alma de jazz."),
    ("Demon Slayer", "anime", 2019, ["Acción","Fantasía"], "Demonios, respiración y redención familiar."),
]

movies = [
    ("The Dark Knight", "movie", 2008, ["Acción","Crimen"], "Batman y el Joker en duelo moral impecable."),
    ("The Wolf of Wall Street", "movie", 2013, ["Comedia","Drama"], "Excesos, dinero y autodestrucción en Wall Street."),
    ("Fight Club", "movie", 1999, ["Drama","Thriller"], "Primera regla: no hablar de Fight Club."),
    ("The Matrix", "movie", 1999, ["Ciencia ficción"], "Pastilla azul o roja; simulación y significado."),
    ("Gladiator", "movie", 2000, ["Acción","Drama"], "Venganza en la Antigua Roma."),
    ("Good Will Hunting", "movie", 1997, ["Drama"], "Genio invisible y terapia humana."),
    ("Joker", "movie", 2019, ["Drama","Crimen"], "Origen trágico del villano sin máscara social."),
    ("1917", "movie", 2019, ["Bélico","Drama"], "Misión en tiempo real en la Primera Guerra Mundial."),
    ("Shutter Island", "movie", 2010, ["Thriller","Misterio"], "Isla, locura y una verdad incómoda."),
    ("Taxi Driver", "movie", 1976, ["Drama","Crimen"], "Noche sucia y soledad en las calles."),
    ("Mad Max: Fury Road", "movie", 2015, ["Acción","Ciencia ficción"], "Persecución implacable en el desierto."),
    ("Ratatouille", "movie", 2007, ["Animación","Comedia"], "Cualquiera puede cocinar con pasión."),
    ("How to Train Your Dragon", "movie", 2010, ["Animación","Aventura"], "Chico, dragón y confianza entre especies."),
    ("Monsters, Inc.", "movie", 2001, ["Animación","Comedia"], "Susto como energía hasta que una niña lo cambia."),
    ("Monsters University", "movie", 2013, ["Animación","Comedia"], "Origen de la sorpresa y la competencia."),
]

already = {x["id"] for x in [
    {"id":"spirited-away"},{"id":"blade-runner-2049"},{"id":"breaking-bad"},
    {"id":"random-access-memories"},{"id":"fmab"},{"id":"parasite"},
    {"id":"the-bear"},{"id":"in-rainbows"}
]}

entries = []

def add(title, kind, year, genres, note):
    sid = slugify(title)
    if sid in already:
        sid = sid + "-2"
        if sid in already:
            sid = sid + "-" + os.urandom(2).hex()
    already.add(sid)
    entries.append({
        "id": sid,
        "type": kind,
        "title": title,
        "creator": "",
        "year": year,
        "rating": 0,
        "genres": genres,
        "note": note,
        "favorite": False,
        "poster": None,
        "added": "2025-07-07",
    })

for t,k,y,g,n in series: add(t,k,y,g,n)
for t,k,y,g,n in anime: add(t,k,y,g,n)
for t,k,y,g,n in movies: add(t,k,y,g,n)

print(f"[prepare] total entries: {len(entries)}")

# Enrich posters
enrich = []
anime_titles = [(e["title"], e["year"]) for e in entries if e["type"] == "anime"]
movie_titles = [(e["title"], e["year"]) for e in entries if e["type"] == "movie"]
series_titles = [(e["title"], e["year"]) for e in entries if e["type"] == "series"]

poster_map = {}

for title, year in anime_titles:
    res = anilist_poster(title, year)
    poster_map[title] = res

for title, year in series_titles:
    res = itunes_artwork(title, media="tvShow", year=year)
    poster_map[title] = res

for title, year in movie_titles:
    res = itunes_artwork(title, media="movie", year=year)
    poster_map[title] = res

missing = []
for e in entries:
    res = poster_map.get(e["title"])
    if res and res.get("poster"):
        e["poster"] = res["poster"]
        if not e["creator"]:
            e["creator"] = ""
        if not e["year"] and res.get("year"):
            e["year"] = res["year"]
    else:
        missing.append(e["title"])

print(f"[posters] with poster: {sum(1 for e in entries if e.get('poster'))}/{len(entries)}")
if missing:
    print(f"[posters] missing: {missing}", file=sys.stderr)

# Save output
with open("data/seed-extra.js", "w", encoding="utf-8") as f:
    f.write("/*\n * seed extra: added series, anime and movies requested by user.\n */\n")
    f.write("window.SEED_EXTRA = ")
    json.dump(entries, f, ensure_ascii=False, indent=2)
    f.write(";\n")

print("[done] wrote data/seed-extra.js")

File: scripts/populate-posters.js
/**
 * populate-posters.js
 *
 * Bulk-fill missing `poster` values for:
 * - anime  -> AniList GraphQL (no key)
 * - movies/series -> TMDB (requires key)
 * - music  -> iTunes Search API (no key)
 *
 * Usage:
 *   node populate-posters.js
 *
 * Writes updated files back to:
 *   data/seeds.js
 *   data/seed-extra.js
 *   data/seeds.js.bak / data/seed-extra.js.bak
 */

const fs = require('fs')
const path = require('path')

const ROOT = path.join(__dirname, '..')
const FILES = [
  path.join(ROOT, 'data/seeds.js'),
  path.join(ROOT, 'data/seed-extra.js')
]

const TMDB_API_KEY_STORAGE = 'emotions_vault_tmdb_key'
const TMDB_BASE = 'https://api.themoviedb.org/3'
const TMDB_IMG_BASE = 'https://image.tmdb.org/t/p'

function esc(str) {
  return String(str)
    .replace(/\\/g, '\\\\')
    .replace(/"/g, '\\"')
    .replace(/\n/g, '\\n')
    .replace(/\r/g, '')
}

async function fetchJSON(url, opts) {
  const res = await fetch(url, opts || {})
  if (!res.ok) throw new Error('HTTP ' + res.status + ' for ' + url)
  return res.json()
}

async function tmdbConfig(key) {
  const data = await fetchJSON(`${TMDB_BASE}/configuration?api_key=${encodeURIComponent(key)}`)
  return data.images || null
}

async function fetchTMDBPoster(title, type, year, key) {
  const mediaType = type === 'series' ? 'tv' : 'movie'
  const params = new URLSearchParams({
    api_key: key,
    query: title,
    include_adult: 'false',
    language: 'es-ES',
    page: '1'
  })
  if (year) params.set('year', String(year))

  const data = await fetchJSON(`${TMDB_BASE}/search/${mediaType}?${params.toString()}`)
  const results = data.results || []
  if (!results.length) return null

  let best = results[0]
  if (year) {
    const withYear = results.find(r => (r.release_date || r.first_air_date || '').startsWith(String(year)))
    if (withYear) best = withYear
  }

  const posterPath = best.poster_path
  if (!posterPath) return null

  const config = await tmdbConfig(key)
  let size = 'w500'
  if (config && config.poster_sizes && config.poster_sizes.includes('w780')) size = 'w780'

  return {
    title: best.title || best.name || title,
    year: (best.release_date || best.first_air_date || '').slice(0, 4) || year,
    poster: `${TMDB_IMG_BASE}/${size}${posterPath}`,
    source: 'tmdb'
  }
}

async function fetchAniListPoster(title, year) {
  const query = `
    query ($search: String) {
      Page(page: 1, perPage: 3) {
        media(search: $search, type: ANIME, sort: POPULARITY_DESC) {
          id
          title { romaji english native }
          coverImage { extraLarge large }
          startDate { year }
        }
      }
    }
  `

  const res = await fetch('https://graphql.anilist.co', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', Accept: 'application/json' },
    body: JSON.stringify({ query, variables: { search: title } })
  })
  if (!res.ok) throw new Error('AniList HTTP ' + res.status)
  const json = await res.json()
  const media = (json.data && json.data.Page && json.data.Page.media && json.data.Page.media[0]) || null
  if (!media) return null

  const cover = media.coverImage && (media.coverImage.extraLarge || media.coverImage.large)
  if (!cover) return null

  return {
    title: media.title && (media.title.romaji || media.title.english || media.title.native || title),
    year: (media.startDate && media.startDate.year) || year,
    poster: cover,
    source: 'anilist'
  }
}

async function fetchiTunesPoster(title, year) {
  const params = new URLSearchParams({
    term: title,
    media: 'music',
    limit: '5',
    country: 'US'
  })
  const data = await fetchJSON(`https://itunes.apple.com/search?${params.toString()}`)
  const results = data.results || []
  if (!results.length) return null

  let best = results[0]
  results.forEach(r => {
    const bestSize = (best.artworkUrl100 || '').replace('100x100', '600x600')
    const candidateSize = (r.artworkUrl100 || '').replace('100x100', '600x600')
    if (candidateSize && (!bestSize || candidateSize.length > bestSize.length)) best = r
  })

  const artwork = (best.artworkUrl100 || '').replace('100x100', '600x600')
  if (!artwork) return null

  return {
    title: best.collectionName || best.trackName || title,
    year: (best.releaseDate || '').slice(0, 4) || year,
    poster: artwork,
    source: 'itunes'
  }
}

async function fetchPoster(title, type, year, tmdbKey) {
  const normalizedType = String(type).toLowerCase()
  if (normalizedType === 'anime') return fetchAniListPoster(title, year)
  if (normalizedType === 'music') return fetchiTunesPoster(title, year)
  if ((normalizedType === 'movie' || normalizedType === 'series') && tmdbKey) return fetchTMDBPoster(title, normalizedType, year, tmdbKey)
  return null
}

async function main() {
  const tmdbKey = process.argv.includes('--tmdb-key') ? process.argv[process.argv.indexOf('--tmdb-key') + 1] : ''
  if (!tmdbKey) {
    // try read from localStorage file if exists
    try {
      const localStoragePath = path.join(process.env.APPDATA || process.env.LOCALAPPDATA || '', '..', 'Local', 'hermes', 'localStorage')
      // No local file to read reliably, continue without key
    } catch (_e) {}
  }

  for (const file of FILES) {
    if (!fs.existsSync(file)) continue

    const original = fs.readFileSync(file, 'utf8')
    if (!original.includes('window.SEED_EXTRA =') && !original.includes('window.SEEDS =')) continue

    const backup = file + '.bak'
    fs.writeFileSync(backup, original)

    const itemsMatch = original.match(/window\.(SEED_EXTRA|SEEDS)\s*=\s*(\[\s*[\s\S]*?\]);/)
    if (!itemsMatch) {
      console.log('No items array found in', file)
      continue
    }

    let items
    try {
      items = JSON.parse(itemsMatch[2])
    } catch (e) {
      console.log('JSON parse error in', file, e.message)
      continue
    }

    let updatedCount = 0
    for (const item of items) {
      if (item.poster) continue
      try {
        const result = await fetchPoster(item.title, item.type, item.year, tmdbKey)
        if (result && result.poster) {
          item.poster = result.poster
          updatedCount++
          console.log(`  ✓ ${item.title} -> ${result.source}`)
        } else {
          console.log(`  ✗ ${item.title} -> no poster found`)
        }
      } catch (e) {
        console.log(`  ! ${item.title} -> error: ${e.message}`)
      }
    }

    const newPayload = itemsMatch[1] + ' = ' + JSON.stringify(items, null, 2) + ';'
    const updated = original.replace(itemsMatch[0], newPayload)
    fs.writeFileSync(file, updated)
    console.log(`Updated ${updatedCount} posters in ${path.basename(file)}`)
  }
}

main().catch(err => {
  console.error(err)
  process.exit(1)
})
